### PR TITLE
feat(types,clerk-js): Add support for backup code operations in UserProfille

### PR DESCRIPTION
**Pull request by chanioxaris**
## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [x] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend-core`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/edge`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

Add the supported operations for backup codes in UserProfile. Users can select and generate a new set of backup codes if not already set and any other multi-factor method is enabled. Can also regenerate to get a new fresh set anytime they want. Users can copy in clipboard, download as txt file and print the codes

### Before
Not supported

### After

https://user-images.githubusercontent.com/22435234/192485087-14380056-fbb8-4570-81ba-d0fb249e9ba5.mov

https://www.notion.so/clerkdev/Introduce-backup-code-operations-in-UserProfile-fdec6cbc1ffc4c5aa66456f3ca348b5e

### CODE REVIEW

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodePage.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+const AddBackupCode = (props: AddBackupCodeProps) => {
+  const { title, onContinue } = props;
+
+  return (
+    <ContentPage.Root headerTitle={title}>
+      <Text>Backup codes will be enabled for this account.</Text>
+      <Text>
+        Keep the backup codes secret and store them securely. You may regenerate backup codes if you think they have
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/UserProfile/MfaBackupCodePage.tsx:38]: 🙃 
```suggestion
        Keep the backup codes secret and store them securely. You may regenerate backup codes if you suspect they have
```

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodePage.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  title: string;
+  onContinue: () => void;
+};
+
+const AddBackupCode = (props: AddBackupCodeProps) => {
+  const { title, onContinue } = props;
+
+  return (
+    <ContentPage.Root headerTitle={title}>
+      <Text>Backup codes will be enabled for this account.</Text>
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/UserProfile/MfaBackupCodePage.tsx:36]: ❓ @nikosdouvlis Should we add all these strings in the new localization context or will we do this in a separate PR?

**Review by SokratisVidros — CHANGES_REQUESTED** [review]
Nicely done @chanioxaris

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodePage.tsx (hunk 2 above)
Comment (nikosdouvlis) [reply packages/clerk-js/src/ui/UserProfile/MfaBackupCodePage.tsx:36]: Hello Sokrati, I've already talked with Haris about that. Lets "hardcode" them for now, and I'll do a final pass to localize all strings under `/UserProfile`

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+type MfaBackupCodeProps = {
+  backupCodes?: string[];
+};
+
+const MfaBackupCode = (props: MfaBackupCodeProps) => {
+  const { backupCodes } = props;
+  const { applicationName } = useEnvironment().displayConfig;
+  const user = useCoreUser();
+  const { print, printableProps } = usePrintable();
+  const { onCopy, hasCopied } = useClipboard(backupCodes?.join(',') as string);
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx:76]:
```suggestion
  const { onCopy, hasCopied } = useClipboard(backupCodes?.join(',') || '');
```

Comment (nikosdouvlis) [reply packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx:76]: That way we can get rid of the manual type assertion and be safe if `backupCodes` end up as undefined for some reason.

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        <Text
+          localizationKey={'Store them securely and keep them secret.'}
+          variant='smallRegular'
+          colorScheme='neutral'
+        />
+      </Col>
+      <Grid
+        gap={2}
+        sx={t => ({
+          gridTemplateColumns: `repeat(5, minmax(${t.sizes.$12}, 1fr))`,
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx:109]: ❓ We'll test more on staging, but does this resolve the responsiveness issue we noticed locally?

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+          variant='link'
+          colorScheme='primary'
+          onClick={print}
+        >
+          Print
+        </Button>
+      </Flex>
+
+      <PrintableComponent {...printableProps}>
+        <Heading>Your backup codes are:</Heading>
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx:150]: 🙃 Should we also include the  application name and possibly an identifier?

File: packages/clerk-js/src/ui/UserProfile/Content.tsx [hunk 6]
```diff
@@ -94,6 +95,9 @@ export const Content = React.forwardRef<HTMLDivElement>((_, ref) => {
             <Route path='totp/remove'>
               <RemoveMfaTOTPPage />
             </Route>
+            <Route path='backup_code/add'>
```
[the remark this answers is not in the document]
Comment (chanioxaris) [reply packages/clerk-js/src/ui/UserProfile/Content.tsx:98]: Not a strong opinion, just went with add as we use it as copyright in our UserProfile screens

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx (hunk 4 above)
Comment (chanioxaris) [reply packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx:109]: Yeap, did some testing and seems to work find in every screen. We will test it more on staging as you said and see if there is any issue to resolve

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx (hunk 5 above)
Comment (chanioxaris) [reply packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx:150]: Added both in the same way we do for the txt file

File: packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+      </FormButtonContainer>
+    </ContentPage.Root>
+  );
+};
+
+type MfaBackupCodeProps = {
+  backupCodes?: string[];
+};
+
+const MfaBackupCode = (props: MfaBackupCodeProps) => {
```
[the remark this answers is not in the document]
Comment (chanioxaris) [reply packages/clerk-js/src/ui/UserProfile/MfaBackupCodeCreatePage.tsx]: Splitted in 3 different files
- `MfaBackupCodeCreatePage.tsx`
- `MfaBackupCodeList.tsx`
- `MfaBackupCodeTile.tsx`

Let me know if this works

**Review by SokratisVidros — APPROVED** [review]

**Review by nikosdouvlis — APPROVED** [review]
